modules/packages/freebsd_ports

```diff
@@ -0,0 +1,108 @@
+#!/bin/sh -e
+
+command=$1
+
+while read -r line; do
+  export INPUT_$line
+done
+
+get_package_data() {
+  name="${INPUT_File?File must be given to get-package-data}"
+  echo PackageType=repo
+  echo Name=$name
+}
+
+list_installed() {
+  # Example pkg output:
+  # sudo-1.8.14p3
+  # Name           : sudo
+  # Version        : 1.8.14p3
+  # Installed on   : Sun Aug 16 05:36:05 UTC 2015
+  # Origin         : security/sudo
+  # Architecture   : freebsd:10:x86:64
+  #
+  # After rewrite:
+  # Name=sudo
+  # Version=1.8.14p3
+  # Architecture=none
+  pkg info -f -a | egrep '^(Name|Version|Architecture)' | sed -e 's/[ ]*:[ ]*/=/' -e 's/^Architecture=.*/Architecture=none/'
+}
+
+repo_install() {
+  name="${INPUT_Name?Name must be given to repo-install}"
+  version="${INPUT_Version}"
+
+  export BATCH=1
+  PORT_PATH=$(whereis -sq "$name")
+
+  if [ -z "$PORT_PATH" ]
+  then
+    echo "ErrorMessage=Could not install $name, port does not exist"
+    exit 0
+  fi
+
+  cd "$PORT_PATH"
+
+  if [ -n "$version" ]
+  then
+    available=$(make -V PORTVERSION)
+    if [ "$available" != "$version" ]
+    then
+      echo "ErrorMessage=Could not install $name $version, available version was $available"
+      exit 0
+    fi
+  fi
+
+  make deinstall reinstall >&2
```

Comment:
In testing on my own machines I found that "make install" failed sometimes (if a package is removed via "pkg del" instead of "make deinstall" then "make install" does nothing), but "make deinstall" doesn't error if there's nothing to be done, so this worked out the most reliable way of installing or upgrading a port.